### DISCUSSION TITLE
Question about final model (& checkpoint averaging)

### DISCUSSION THREAD

**Discussion by pehonnet** [NVIDIA-NeMo/NeMo#5150] [Q&A] [answered]
Hi, thanks for the great toolkit.
I have a question regarding how a final `.nemo` file is produced when training. The reason I am asking is I have tested a model trained on my data with different settings:
- the final `.nemo` model produced by `speech_to_label.py` script
- the averaged `.nemo` model using the top 3 `.ckpt` models
- each of the top 3 `.ckpt` models converted to `.nemo`
and I don't get the same results for any of these, meaning that the *final* `.nemo` model is neither the best `.ckpt` nor the average of the top 3 `.ckpt`.
Can you confirm this finding? And could you explain how the final `.nemo` model is produced? Thanks!

**Comment by titu1994** [comment, marked as answer]
The final Nemo file produced by speech to label is simply the model at the end of training the last step - without averaging. One of the 3 ckpt files should be called ***-last.ckpt - this is what's used to build the last Nemo file.

**Reply by pehonnet** [reply]
Thank you for the answer. Then it is strange as I am getting different results when using the checkpoint without `last` in the name compared to final and also compared to the same one without `final` in the name.

It also looks like they are not exactly the same file:
```bash
$ shasum mymodel--val_loss=0.1465-epoch=199.ckpt
e45062bb1c69bda5375d6edbccd07086b85a34b3  mymodel--val_loss=0.1465-epoch=199.ckpt
$ shasum mymodel--val_loss=0.1465-epoch=199-last.ckpt
681f6d85719ca4ed79e71fd1b239413773a20b40  mymodel--val_loss=0.1465-epoch=199-last.ckpt
```
Some detailed results:
| Model | accuracy |
|-- | --|
| final (`.nemo`)| 98.05% |
| average (155, 180, 199) | 97.54% |
| epoch 155 (0.1458) | 97.37% |
| epoch 180 (0.1435) | 97.42% |
| epoch 199 (0.1465) | 98.03% |
| epoch 199-last (0.1465) | 98.08% |

How is the `-last.ckpt` file produced? It looks like it is not simply copying the file from the last epoch. Otherwise maybe I misunderstand and the file `epoch=199` is actually the file before running epoch 199, and `epoch=199-last` is after running epoch 199?

Another point is, to be able to run my inference script I convert the ckpt to nemo format, but I have not actually written a script to convert them, I used the model averaging script, with only 1 model, so I would expect it to stay the same from what I see in the model averaging code.

EDIT: the final was not 98.18% but 98.05%.

**Reply by titu1994** [reply]
This is what gets called on end of training to save the  Nemo file. 

https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/exp_manager.py#L818

**Reply by pehonnet** [reply]
Thanks! So it means that there could be slight difference because some of the checkpoints are on full epoch and some are on "interval" ~ iterations?

**Reply by titu1994** [reply]
There's no interval ones. We only save on validation.
